Add cold/offline backup script (#4459)
strimzi/strimzi-kafka-operator: tools/cold-backup/README.md

# Strimzi backup
Bash script for cold/offline backups of Kafka clusters on Kubernetes/OpenShift.

> If you think you do not need a backup strategy for Kafka as it has embedded data replication,

Reviewer: ```suggestion
If you think you do not need a backup strategy for Kafka because of its embedded data replication,
```

Author: Thanks :)

> then try to immagine a misconfiguration/bug/security-breach deleting all your data. For hot/online

Reviewer: ```suggestion
then consider the impact of a bug or security breach that deleting all your data. For hot/online
```

Author: Thanks :)

> then try to immagine a misconfiguration/bug/security-breach deleting all your data. For hot/online

Author: OK

> backups, you should probably look at storage snapshotting or streaming into object storage.

Reviewer: ```suggestion
backups, you can use storage snapshotting or streaming into object storage.
```

Author: Thanks :)

> To run the script, the K8s user must have rights to work with PVC and use Strimzi custom resources.

Reviewer: ```suggestion
To run the script, the Kubernetes user must have permission to work with PVC and Strimzi custom resources.
```

Author: Thanks :)

> The backup procedure will stop operator and selected cluster for the entire duration. Before restoring
> the Kafka cluster you need to make sure to have the right version of Strimzi CRDs installed. If you
> have a single cluster-wide operator, then you need to scale it down manually. You can run backup and
> restore procedures of different Kafka clusters in parallel. Only local file system is supported.
> ConsumerGroup offsets are included, but not KafkaConnect, MirrorMaker and KafkaBridge CRs.

Reviewer: ```suggestion
The procedure will stop the Cluster Operator and selected cluster for the duration of the backup. Before restoring
the Kafka cluster you need to make sure to have the right version of Strimzi CRDs installed. If you
have a single cluster-wide Cluster Operator, then you need to scale it down manually. You can run backup and
restore procedures for different Kafka clusters in parallel. Only the local file system is supported.
`ConsumerGroup` offsets are included, but not Kafka Connect, MirrorMaker and Kafka Bridge custom resources.
```

Author: Thanks :)

# Strimzi backup
Bash script for cold/offline backups of *Strimzi* clusters on *Kubernetes/OpenShift*.

If you think you do not need a backup strategy for Kafka as it has embedded data replication,
then try to immagine a misconfiguration/bug/security-breach deleting all your data. For hot/online
backups, you should look at *MirrorMaker2* to sync with a remote cluster, but this comes with
additional complexities and required resources.

> To run the script you must be logged in as *cluster-admin* user. The script will stop the operator

Reviewer 2: Does it really need to be _cluster-admin_ user? I don't see anything what would require it. It might be better to list the permissions the user needs (not necessarily listing every RBAC, but for example _you need to be abel to work with PVC and use the Strimzi custom resources_ etc.)

Author: Right. Let me double check this.

and the whole cluster for the entire duration of the backup procedure. If you have a single cluster
wide operator, then you need to manually scale it down before starting the backup.

The final archive contains an *env* file with the operator's version that you need to deploy *after*
the restore ahs finished. Only local file system is supported, consumer group offsets are included,
but not KafkaConnect custom images, that are usually hosted on an external registry.

## Requirements
- bash 5+ (GNU)

> - kubectl 1.20+ (Kubernetes)

Author: Probably not. I was just using the one provided by `brew`. Let me do another test and change it to the supported one.

> - kubectl 1.20+ (Kubernetes)

Reviewer 3: I guess if STs pass, we can just change the requirement to 1.16+

> - kubectl 1.20+ (Kubernetes)

Author: Hi @Frawless , I just pushed a new commit tested with 1.16. I also added a 5s sleep before running the backup script, in order to make the test more robust.

    --producer-props acks=1 bootstrap.servers=my-cluster-kafka-bootstrap:9092

kubectl exec -it my-cluster-kafka-0 -c kafka -- \
    bin/kafka-console-consumer.sh --bootstrap-server :9092 \
    --topic my-topic --group my-group --from-beginning --timeout-ms 15000

# save consumer group offsets
kubectl exec -it my-cluster-kafka-0 -c kafka -- \
    bin/kafka-consumer-groups.sh --bootstrap-server :9092 \
    --group my-group --describe > /tmp/offsets.txt

# send additional 12345 messages
kubectl run kafka-producer-perf-test -it \
    --image="quay.io/strimzi/kafka:latest-kafka-2.6.0" \
    --rm="true" --restart="Never" -- bin/kafka-producer-perf-test.sh \
    --topic my-topic --record-size 1000 --num-records 12345 --throughput -1 \
    --producer-props acks=1 bootstrap.servers=my-cluster-kafka-bootstrap:9092

# run backup procedure
./tools/cold-backup/run.sh -b -n myproject -c my-cluster -t /tmp/my-cluster.zip -m custom-cm

# delete the namespace and restore
kubectl delete ns myproject
kubectl create ns myproject
./tools/cold-backup/run.sh -r -n myproject -c my-cluster -s /tmp/my-cluster.zip

# deploy the operator and wait for provisionig

> kubectl create -f ./install/cluster-operator

Reviewer 2: Who and where installs the Kafka custom resource? I assume the backup/restore tool does it? You should then probably install the operator first before running the restore. In case you would be restoring on a different cluster, you might not have the CRDs installed until this step and that would fail the restore.

Author: Yes, I thought about this, but in practice doing like this ends with zookeeper error on TLS certificates. More details later.

Author: This is the issue with switching these two steps in the ST. It seem to work when doing it manually. Will do some more tests.

```
2021-03-16 15:26:05,574 WARN Exception caught (org.apache.zookeeper.server.NettyServerCnxnFactory) [nioEventLoopGroup-7-2]
io.netty.handler.codec.DecoderException: javax.net.ssl.SSLHandshakeException: PKIX path validation failed: java.security.cert.CertPathValidatorException: signature check failed
	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:471)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:276)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:163)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:714)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:650)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:576)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:493)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

Author: Ok I fixed that and switched the operator deploy before restore procedure.
